crypto/bio/cipher.c

```diff
@@ -0,0 +1,300 @@
+// Copyright Amazon.com, Inc. or its affiliates. All Rights Reserved.
+// SPDX-License-Identifier: Apache-2.0 OR ISC
+
+#include <errno.h>
+#include <openssl/buffer.h>
+#include <openssl/crypto.h>
+#include <openssl/evp.h>
+#include <openssl/mem.h>
+#include <openssl/pkcs7.h>
+#include <stdio.h>
+#include "../crypto/bio/internal.h"
+#include "../fipsmodule/cipher/internal.h"
+#include "../internal.h"
+#include "./internal.h"
+
+static int enc_write(BIO *h, const char *buf, int num);
+static int enc_read(BIO *h, char *buf, int size);
+static long enc_ctrl(BIO *h, int cmd, long arg1, void *arg2);
+static int enc_new(BIO *h);
+static int enc_free(BIO *data);
+#define ENC_BLOCK_SIZE (1024 * 4)
+
+typedef struct enc_struct {
+  uint8_t done;     // > 0 when finished
+  uint8_t flushed;  // >0 when buffered data has been flushed
+  uint8_t ok;       // decryption status
+  int buf_off;      // start idx of buffered data
+  int buf_len;      // length of buffered data
+  EVP_CIPHER_CTX *cipher;
+  uint8_t buf[ENC_BLOCK_SIZE];
+} BIO_ENC_CTX;
+
+static const BIO_METHOD methods_enc = {
+    BIO_TYPE_CIPHER,
+    "cipher",
+    enc_write,
+    enc_read,
+    NULL,  // enc_puts
+    NULL,  // enc_gets
+    enc_ctrl,
+    enc_new,
+    enc_free,
+    NULL,  // enc_callback_ctrl
+};
+
+const BIO_METHOD *BIO_f_cipher(void) { return &methods_enc; }
+
+static int enc_new(BIO *bi) {
+  BIO_ENC_CTX *ctx;
+
+  if ((ctx = OPENSSL_zalloc(sizeof(*ctx))) == NULL) {
+    return 0;
+  }
+
+  ctx->cipher = EVP_CIPHER_CTX_new();
+  if (ctx->cipher == NULL) {
+    OPENSSL_free(ctx);
+    return 0;
+  }
+  ctx->done = 0;
+  ctx->flushed = 0;
+  ctx->ok = 1;
+  ctx->buf_off = 0;
+  ctx->buf_len = 0;
+  BIO_set_data(bi, ctx);
+  BIO_set_init(bi, 1);
+
+  return 1;
+}
+
+static int enc_free(BIO *a) {
+  BIO_ENC_CTX *b;
+
+  if (a == NULL) {
+    return 0;
+  }
+
+  b = BIO_get_data(a);
+  if (b == NULL) {
+    return 0;
+  }
+
+  EVP_CIPHER_CTX_free(b->cipher);
+  OPENSSL_cleanse(b, sizeof(BIO_ENC_CTX));
+  OPENSSL_free(b);
+  BIO_set_data(a, NULL);
+  BIO_set_init(a, 0);
+
+  return 1;
+}
+
+static int enc_read(BIO *b, char *out, int outl) {
+  BIO_ENC_CTX *ctx = BIO_get_data(b);
+  if (ctx == NULL) {
+    return 0;
+  }
+  BIO *next = BIO_next(b);
+  if (next == NULL) {
+    return 0;
+  }
+
+  int bytes_processed = 0;
+  int remaining = outl;
+  while (remaining > 0) {
+    if (ctx->buf_len > 0) {
+      int bytes_decrypted;
+      int to_decrypt = remaining > ctx->buf_len ? ctx->buf_len : remaining;
+      char *out_pos = out + bytes_processed;
+      if (!EVP_DecryptUpdate(ctx->cipher, (uint8_t *)out_pos, &bytes_decrypted,
```

Comment:
The condition on this call to `EVP_DecryptUpdate` needs to be more strict. From its documentation:
> The number of output bytes may be up to |in_len| plus the block length minus one and |out| must have sufficient space.

So it must be the case that: `remaining > (to_decrypt + ctx->block_len - 1)`

Reply:
good point. i've added this guard, slightly re-written for clarity.